### PR TITLE
[data] Pre-reqs for implementing stage fusion

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This implements a bunch of the mechanical changes needed for https://github.com/ray-project/ray/pull/22373. It's split out from that so that the core changes can be reviewed separately.

### CODE REVIEW

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:482]: Moved to separate file for clarity.

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:466]: Add the concept of a block_udf which is later used to implement fusion of normal and map shuffle stages.

File: python/ray/data/impl/fast_repartition.py
Comment (ericl) [inline python/ray/data/impl/fast_repartition.py:15]: Moved verbatim except for this `wrapped_ds` line.

File: python/ray/data/impl/shuffle.py
Comment (clarkzinzow) [inline python/ray/data/impl/shuffle.py:124]: Nit: Found the `pieces` naming to be odd.
```suggestion
        blocks = list(block_udf(block))
        if len(blocks) > 1:
            builder = BlockAccessor.for_block(blocks[0]).builder()
            for b in blocks:
                builder.add_block(b)
            block = builder.build()
        else:
            block = blocks[0]
```

File: python/ray/data/impl/sort.py
Comment (clarkzinzow) [inline python/ray/data/impl/sort.py:109]: I was looking at our sort last night and noticed that we weren't releasing the input blocks after starting the map _and_ the map results after starting the reduce/merge. It'd be nice if we could release the latter as well down below:
```python
del map_results
merge_bar = ProgressBar("Sort Merge", len(reduce_results))
merge_bar.block_until_complete([ret[0] for ret in reduce_results])
```

This can wait for a follow-up PR, if you'd like.

**Review by clarkzinzow — APPROVED** [review]
LGTM, both review items/nits can wait for a follow-up PR, so feel free to merge!

File: python/ray/data/impl/shuffle.py
Comment (ericl) [reply python/ray/data/impl/shuffle.py:124]: Thanks. I'll fix this in the follow-up

File: python/ray/data/impl/sort.py
Comment (ericl) [reply python/ray/data/impl/sort.py:109]: Good point. I'll make a mini follow-up including this.

**Review by scv119 — APPROVED** [review]

**Pull request merged**
